Enables inhibitory stdp

weidel-p opened the pull request:
With this pull request, STDP is possible for inhibitory connections by assigning negative values for `weight` and `Wmax` in STDPConnection and STDPTripletConnection.
The functions `facilitate` and `depress` assumes positive weight changes, but the actual sign of the weight change is applied in the `send` function.
The sign of `weight` determines the sign of `Wmax` during connection and SetStatus. During simulation, the sign of `Wmax` determines the sign of `weight`.

This addresses #240.

weidel-p commented:
I suggest @abigailm, @mhelias and/or @suku248 as reviewers.

heplesser commented on models/stdp_connection.h:
```diff
@@ -185,7 +187,7 @@ class STDPConnection : public Connection< targetidentifierT >
     double_t norm_w =
       ( w / Wmax_ ) - ( alpha_ * lambda_ * std::pow( w / Wmax_, mu_minus_ ) * kminus );
-    return norm_w > 0.0 ? norm_w * Wmax_ : 0.0;
+    return norm_w > 0.0 ? norm_w * Wmax_ : copysign( 0.0, Wmax_ );
```
`copysign( 0.0, Wmax_ )` returns a value with the magnitude of its first and the sign of its second argument. So here we would get +0 og -0. Does that make sense?

weidel-p replied:
With this line, I wanted to make sure that the sign of `weight` does not flip from negative to positive if it becomes 0. After having another look into this and after testing, I agree that it was unnecessary.
I fixed that in 0ebe4ba3d9d112bf1af7d999f3bd3c0f1789b0ad.

heplesser commented:
@weidel-p The `copysign()` function is part of C++11. Since we cannot rely on C++11 support on all relevant compilers, we need a configure test for whether `copysign()` is available and if not, we need to define it ourselves, e.g., in `libnestutil.numerics.h`.

weidel-p commented:
@heplesser  `copysign()` is also available in C99. In 0ebe4ba3d9d112bf1af7d999f3bd3c0f1789b0ad I changed the include to use the C header. Is this already sufficient to make sure all relevant compilers support this or shall I include the function to `libnestutil.numerics.h` as you suggested?

heplesser commented:
@weidel-p I tried
```
#include <math.h>
#include <iostream>

int main()
{
  double x = 10;
  double y = -2;

  double z = copysign(x, y);

  std::cout << x << '\t' << y << '\t'  << z << std::endl;
  
  return 0;
}
```
using xlc++, Intel, Portland and Clang compilers. It works everywhere. The only system I cannot test is K. Could you ask our K-colleagues to test it?

jakobj commented:
with FCCpx and mpiFCCpx @heplesser's script works also on K. no compiler warnings/errors and the output reads `10      -2      -10` which is correct as far as I can see.

heplesser commented:
@jakobj Thanks for checking on K!
@weidel-p Could you make a tiny, inconsequential change to trigger a new build? The last build failed partially due to time-outs. Conditional on all test passing :+1: from my side, but one of the plasticity experts should approve before merging.

sanjayankur31 commented:
Just a thought - Is using copysign to copy the sign of weight to Wmax under the hood a good idea? Wouldn't it be better to have a bad parameter check (or something) that checks if the two signs are same instead - let users manually set both the signs so that they're aware of exactly what they're doing?

sanjayankur31 commented on models/stdp_triplet_connection.h:
```diff
@@ -185,7 +187,7 @@ class STDPTripletConnection : public Connection< targetidentifierT >
   {
     double_t new_w = w + kplus * ( Aplus_ + Aplus_triplet_ * ky );
-    return new_w < Wmax_ ? new_w : Wmax_;
+    return new_w < std::abs( Wmax_ ) ? new_w : std::abs( Wmax_ );
```
Won't implementation along these lines permit sign switching? As in an excitatory connection reducing, going negative and turning into an inhibitory one and vice versa? 

Example - start with 5,500 - depress and turn into -5, 500 -> check still passes?
Example - start with -5, -500 - facilitate and turn into 5,-500 -> check still passes?

heplesser replied:
Looking at this code again after a comment from @sanjayankur31 which is strangely missing from Github, I find the code very difficult to read. The problem is that on l. 254f you call call `facilitate_()` with the absolute value of `w`, something that is not clear from the code here. Basically, you mix handling of signs in `send()` and `facilitate_()`. I would suggest to do all signe handling inside `facilitate_()`, so that the code in `send()` would again become
```
weight_ = facilitate_( weight_, Kplus_ * std::exp( minus_dt / tau_plus_ ), ky );
```
That way, all sign handling will be inside `facilitate_()`, which we then could code like this
```
const double_t new_w_abs = std::abs( w ) + kplus * ( Aplus_ + Aplus_triplet_ * ky );
return copysign( std::min( new_w_abs, std::abs( Wmax_ ) ), Wmax_ );
```
I suppose the same applies to `depress_()`.

sanjayankur31 replied:
Oh, sorry - I removed the comment - I wasn't quite sure if what I was asking was correct - hadn't worked it out on paper yet.

flinz replied:
@weidel-p I agree with @heplesser and @sanjayankur31 that the handling of the sign issues for weight updates should be put into ``facilitate`` and ``depress`` respectively.

heplesser commented:
@sanjayankur31 I suppose in your comment on changing the sign under the hood you refer to `STDPConnection::set_weight()` and `STDPConnection::set_status()`. I agree that it would make sense to introduce a test in both places, raising an exception if one tries to set a weight (or weight limit) that is not consistent with the sign. Zero should in this case work for both signs. There should also be a test ensuring that this check is in place.

This would stop users from subtle and hard-to-find errors.

tammoippen commented:
@flinz Can you please look into this PR? Thank you.

flinz commented on pynest/nest/tests/test_stdp_triplet_synapse.py:
```diff
@@ -84,11 +85,11 @@ def decay(self, time, Kplus, Kplus_triplet, Kminus, Kminus_triplet):
     def facilitate(self, w, Kplus, Kminus_triplet):
         """Facilitate weight."""
-        return w + Kplus * (self.syn_spec["Aplus"] + self.syn_spec["Aplus_triplet"] * Kminus_triplet)
+        return abs(w + Kplus * (self.syn_spec["Aplus"] + self.syn_spec["Aplus_triplet"] * Kminus_triplet))
```
After incorporating the changes proposed by @heplesser in https://github.com/nest/nest-simulator/pull/284#discussion_r58529377 the test functions for ``facilitate`` and ``depress``, and calls to them, should be adapted accordingly.

flinz commented:
After the line comments are addressed, 👍 from me.

heplesser commented:
@weidel-p #391 is merged and Travis works again. Could you merge all changes from master into your PR so that Travis hopefully turns green again?

weidel-p commented:
Hopefully I addressed all comments with the latest commit. 

1. The signs are now exclusively handled in the `facilitate` and `depress` functions.
2. An exception is thrown if `Wmax` and `weight` are initialized with different signs. This ensures that users get an immediate feedback if `weight` and `Wmax` are set inconsistently.
3. If the issue solved by #333 is reintroduced again, the tests for #284 will fail.

heplesser commented on models/stdp_connection.h:
```diff
@@ -68,11 +68,12 @@
 
 // C++ includes:
-#include <cmath>
+#include <math.h>
```
Comment why you include `math.h`.

weidel-p replied:
We were afraid that `copysign` is not available in the c++ header for some supercomputers. Therefore, I changed it to the C header and we tested this for several compilers. Please read the comments above.

heplesser replied:
@weidel-p Yes, I know. But that information should be here in the C++ file, so that if someone reads this file five years from now will understand. E.g. `// C-header since copysign() is in C99 but not C++98`.

weidel-p replied:
Right, I added this comment to the cpp file in the `stdp_triplet_connection.h`.

heplesser commented on models/stdp_triplet_connection.h:
```diff
@@ -76,10 +76,11 @@
 */
 
-#include <cmath>
+#include <math.h>
```
Comment why you include `math.h`.

weidel-p replied:
see above.

heplesser commented on testsuite/unittests/test_inh_stdp_synapse.sli:
```diff
@@ -0,0 +1,245 @@
+}
+loop
+
+%final_weight w w_max mul sub abs dup = 1e-13 leq =
```
Remove this commented-out line.

heplesser commented on testsuite/unittests/test_inh_stdp_synapse.sli:
```diff
@@ -0,0 +1,245 @@
+loop
+
+%final_weight w w_max mul sub abs dup = 1e-13 leq =
+final_weight w w_max mul sub abs 1e-13 leq assert_or_die
```
It would be more in keeping the testsuite style to put the entire loop from l 185 on up to and including the `leq` into a block and the place the `assert_or_die` after the block. Also, there is a SLI function `ToUnittestPrecision` (or similar) that allows testing with a prescribed number of decimals.

weidel-p replied:
This test is basically copied from `test_stdp_synapse` and adapted for negative signs. Therefore, I tried to modify both tests according to your comments. I am now using `ToUnittestPrecision`  and used a block for the loop. You didn't specify where to put the loop, so I left it where it is.

heplesser commented:
@weidel-p I fear that in all the merging, the sign-handling has disappeared from `stdp_connection.h` again, at least there is no trace of `copysign()` left in that file. Could you double-check this?

weidel-p commented:
@heplesser in `stdp_connection.h` no `copysign()` is needed. In `facilitate` and `depress`, the `weight` is normalized between 0 and 1 and afterwards scaled with `Wmax`. This works for both, negative and positive weights.
As stated in the initial post, this PR addresses #240. All work done on the `stdp_connection.h` is just ensuring consistent behavior between the `stdp_triplet_synapse` and `stdp_synapse` (throwing the same exception).

heplesser commented:
@weidel-p Okay, I overlooked that `stdp_connection` doesn't need `copysign()`. But why does it then include `math.h` instead of `cmath`?

heplesser commented:
@weidel-p The tests look fine now.

weidel-p commented:
@heplesser Sorry, I used `copysign()` in `stdp_connection.h` in an earlier version and forgot to change the include back to `cmath`.

heplesser commented:
:+1: and merging. @weidel-p Thanks for you effort!
